mutiny-core/src/logging.rs

```diff
@@ -30,3 +119,156 @@ impl Logger for MutinyLogger {
         }
     }
 }
+
+async fn build_logging_database() -> Result<Rexie, MutinyError> {
+    // Create a new database
+    let rexie = Rexie::builder(LOGGING_DATABASE_NAME)
+        .version(1)
+        .add_object_store(ObjectStore::new(LOGGING_OBJECT_STORE_NAME))
+        // Build the database
+        .build()
+        .await?;
+
+    Ok(rexie)
+}
+
+#[cfg(any(test, feature = "test-utils"))]
+pub(crate) async fn clear() -> Result<(), MutinyError> {
+    let indexed_db = build_logging_database().await?;
+    let tx = indexed_db.transaction(&[LOGGING_OBJECT_STORE_NAME], TransactionMode::ReadWrite)?;
+    let store = tx.store(LOGGING_OBJECT_STORE_NAME)?;
+
+    store.clear().await?;
+
+    tx.done().await?;
+
+    Ok(())
+}
+
+async fn get_logging_data(rexie: &Rexie) -> Result<Option<Vec<String>>, MutinyError> {
+    // Create a new read-only transaction
+    let transaction = rexie.transaction(&[LOGGING_OBJECT_STORE_NAME], TransactionMode::ReadOnly)?;
+
+    let mut store = transaction.store(LOGGING_OBJECT_STORE_NAME)?;
+
+    let res = get_logging_data_with_transaction_store(&mut store).await;
+
+    transaction.done().await?;
+
+    res
+}
+
+/// Pass in a transaction to get the data for.
+/// It is up to the caller to call `transaction.done()`
+async fn get_logging_data_with_transaction_store(
+    store: &mut rexie::Store,
+) -> Result<Option<Vec<String>>, MutinyError> {
+    let logging_js = store.get(&JsValue::from(LOGGING_KEY)).await?;
+
+    // If the key doesn't exist, we return None
+    if logging_js.is_null() || logging_js.is_undefined() {
+        return Ok(None);
+    }
+
+    let logging_data: Vec<String> = logging_js.into_serde()?;
+
+    Ok(Some(logging_data))
+}
+
+async fn write_logging_data(
+    rexie: &Rexie,
+    recent_logs: &mut Vec<String>,
+) -> Result<(), MutinyError> {
+    // Create a new read-write transaction
+    let transaction =
+        rexie.transaction(&[LOGGING_OBJECT_STORE_NAME], TransactionMode::ReadWrite)?;
+
+    let mut store = transaction.store(LOGGING_OBJECT_STORE_NAME)?;
+
+    // get the existing data so we can append to it, trimming if needed
+    let mut existing_logs = get_logging_data_with_transaction_store(&mut store)
+        .await?
+        .unwrap_or_default();
+    existing_logs.append(recent_logs);
+    if existing_logs.len() > 1_000 {
+        let start_index = existing_logs.len() - 1000;
```

Comment:
should make `1000` a const at the top so it is easy to reference and gives the number more context